typing/typedecl_unboxed.ml

```diff
@@ -16,38 +16,90 @@
 
 open Types
 
+module Callstack = struct
+  type t = Path.t list
+
+  module TypeMap = Btype.TypeMap
+  type map = t TypeMap.t
+
+  let visit p callstack : t =
+    p :: callstack
+
+  let visited p callstack =
+    List.exists (Path.same p) callstack
+
+  let head (callstacks : map) ty =
+    TypeMap.find ty callstacks
+
+  let rec extend callstacks ty new_callstack =
+    if TypeMap.mem ty callstacks then callstacks
+    else
+      let callstacks = TypeMap.add ty new_callstack callstacks in
+      Btype.fold_type_expr (fun callstacks ty' ->
+        extend callstacks ty' new_callstack
+      ) callstacks ty
+
+  let fill ty callstack = extend TypeMap.empty ty callstack
+
+  let expand_head_opt env ty callstacks =
+    let head_callstack = head callstacks ty in
+    let ty = Ctype.expand_head_opt env ty in
+    ty, extend callstacks ty head_callstack
+
+  let apply env params (ty, callstacks) args current_callstack =
+    let ty = Ctype.apply env params ty args in
+    (ty, extend callstacks ty current_callstack)
+end
+
 type t =
   | Unavailable
   | This of type_expr
   | Only_on_64_bits of type_expr
 
+let check_annotated ty callstacks =
```

Comment:
Is this function just checking that `extend_callstacks` works, using logic almost identical to `extend_callstacks`. If so, I'm not sure it is gaining you anything.

Reply:
Yes, this function is used for debugging only. It proved tricky to get an implementation right because of the reliance on type node identity, interleaved with expansions (in particular the call to `expand_head_opt`, which transforms the type graph). It will be possible to remove it if people agree on a final implementation.